library/std/src/sys/unix/fs/dir_fd.rs

```diff
@@ -0,0 +1,385 @@
+        })?;
+        // SAFETY: file descriptor was opened in this fn
+        Ok(unsafe { OwnedFd::from_raw_fd(fd) })
+    }
+
+    #[cfg(any(
+        target_os = "solaris",
+        target_os = "illumos",
+        target_os = "haiku",
+        target_os = "vxworks",
+    ))]
+    fn is_dir(_ent: &DirEntry) -> Option<bool> {
+        None
+    }
+
+    #[cfg(not(any(
+        target_os = "solaris",
+        target_os = "illumos",
+        target_os = "haiku",
+        target_os = "vxworks",
+    )))]
+    fn is_dir(ent: &DirEntry) -> Option<bool> {
+        match ent.entry.d_type {
+            libc::DT_UNKNOWN => None,
+            libc::DT_DIR => Some(true),
+            _ => Some(false),
+        }
+    }
+
+    enum LazyReadDir<'a> {
+        Fd(Option<OwnedFd>), // only Fd(None) temporarily in ensure_open()
+        OpenReadDir(ReadDir, BorrowedFd<'a>), // also store the fd to avoid having to call dirfd(3)
+    }
+
+    impl LazyReadDir<'_> {
+        fn open(path: &CStr) -> io::Result<Self> {
+            let fd = openat_nofollow_dironly(None, path)?;
+            Ok(LazyReadDir::Fd(Some(fd)))
+        }
+
+        fn open_subdir_or_unlink_non_dir(&self, child_name: &CStr) -> io::Result<Option<Self>> {
+            let fd = match openat_nofollow_dironly(Some(self.as_fd()), child_name) {
+                Ok(fd) => fd,
+                Err(err) if matches!(err.raw_os_error(), Some(libc::ENOTDIR | libc::ELOOP)) => {
+                    // not a directory - unlink and return
+                    // (for symlinks, older Linux kernels may return ELOOP instead of ENOTDIR)
+                    cvt(unsafe { unlinkat(self.as_fd().as_raw_fd(), child_name.as_ptr(), 0) })?;
+                    return Ok(None);
+                }
+                Err(err) => return Err(err),
+            };
+            Ok(Some(LazyReadDir::Fd(Some(fd))))
+        }
+
+        fn get_parent(&self) -> io::Result<Self> {
+            let fd = openat_nofollow_dironly(Some(self.as_fd()), unsafe {
+                CStr::from_bytes_with_nul_unchecked(b"..\0")
+            })?;
+            Ok(LazyReadDir::Fd(Some(fd)))
+        }
+
+        fn ensure_open(&mut self) -> io::Result<()> {
+            if let LazyReadDir::Fd(fd_opt) = self {
+                let fd = fd_opt.take().unwrap();
+                let ptr = cvt_p(unsafe { fdopendir(fd.as_raw_fd()) });
+                if let Err(err) = ptr {
+                    *fd_opt = Some(fd); // put the fd back
+                    return Err(err);
+                }
+                let dirp = Dir(ptr?);
+                // fd is now owned by dirp which closes it on drop, so give up ownership
+                let fd = fd.into_raw_fd();
+                // SAFETY: the dirp fd stays valid until self is dropped
+                let fd = unsafe { BorrowedFd::borrow_raw(fd) };
+                // a valid root path is not needed because we do not call any functions
+                // involving the full path of the DirEntrys.
```

Comment:
This should also go as a doc comment on the `OpenReadDir` variant since we're abusing the type a little here.

Reply:
Done.